app/com/linkedin/drelephant/spark/heuristics/StagesAnalyzer.scala

```diff
@@ -0,0 +1,479 @@
+  * @param numTasksWithContainerKilled total number of tasks failed due to the container being
+  *                                    killed by YARN for exceeding memory limits.
+  * @param stageDuration wall clock time for stage to run in ms
+  * @param details information and recommendations
+  */
+private[heuristics] case class StageAnalysis(
+    stageId: Int,
+    rawSpillSeverity: Severity,
+    executionSpillSeverity: Severity,
+    longTaskSeverity: Severity,
+    rawSkewSeverity: Severity,
+    taskSkewSeverity: Severity,
+    failedWithOOMSeverity: Severity,
+    failedWithContainerKilledSeverity: Severity,
+    taskFailureSeverity: Severity,
+    stageFailureSeverity: Severity,
+    gcSeverity: Severity,
+    numTasks: Int,
+    medianRunTime: Option[Double],
+    maxRunTime: Option[Double],
+    memoryBytesSpilled: Long,
+    maxTaskBytesSpilled: Long,
+    inputBytes: Long,
+    numFailedTasks: Int,
+    numTasksWithOOM: Int,
+    numTasksWithContainerKilled: Int,
+    stageDuration: Option[Long],
+    details: Seq[String])
+
+/**
+  * Analyzes the stage level metrics for the given application.
+  *
+  * @param heuristicConfigurationData heuristic configuration data
+  * @param data Spark application data
+  */
+private[heuristics] class StagesAnalyzer(
+    private val heuristicConfigurationData: HeuristicConfigurationData,
+    private val data: SparkApplicationData) {
+
+  import ConfigurationUtils._
+
+  // serverity thresholds for execution memory spill
+  val executionMemorySpillThresholds: SeverityThresholds =
+    SeverityThresholds.parse(heuristicConfigurationData.getParamMap
+      .get(SPARK_STAGE_EXECUTION_MEMORY_SPILL_THRESHOLD_KEY), ascending = true)
+      .getOrElse(DEFAULT_EXECUTION_MEMORY_SPILL_THRESHOLDS)
+
+  // severity thresholds for task skew
+  val taskSkewThresholds: SeverityThresholds =
+    SeverityThresholds.parse(heuristicConfigurationData.getParamMap
+      .get(SPARK_STAGE_TASK_SKEW_THRESHOLD_KEY), ascending = true)
+      .getOrElse(DEFAULT_TASK_SKEW_THRESHOLDS)
+
+  // severity thresholds for task duration (long running tasks)
+  val taskDurationThresholds: SeverityThresholds =
+    SeverityThresholds.parse(heuristicConfigurationData.getParamMap
+      .get(SPARK_STAGE_TASK_DURATION_THRESHOLD_KEY), ascending = true)
+      .getOrElse(DEFAULT_TASK_DURATION_THRESHOLDS)
+
+  /**
+    * Get the analysis for each stage of the application
+    * @param curNumPartitions the configured number of partitions for the application
+    *                         (value of spark.sql.shuffle.partitions).
+    * @return list of analysis results of stages.
+    */
+  def getStageAnalysis(curNumPartitions: Int): Seq[StageAnalysis] = {
+    data.stagesWithFailedTasks.map { stageData =>
+      (stageData.stageId, stageData.tasks.map(tasks => tasks.values))
+    }.toMap
+
+    val failedTasksStageMap = data.stagesWithFailedTasks.map { stageData =>
```

Comment:
can use flatMap here instead of map and then flatten.

Reply:
Thanks, changed to flatMap.